File: app/src/main/kotlin/com/openclaw/ghostcrab/domain/model/SkillInstallError.kt
package com.openclaw.ghostcrab.domain.model

sealed class SkillInstallError(val isRetryable: Boolean) {
    data class Unauthorized(val missingScope: String) : SkillInstallError(false)
    data class NotFound(val slug: String) : SkillInstallError(false)
    data class DependencyConflict(val conflicts: List<String>) : SkillInstallError(false)
    data class Network(val cause: String) : SkillInstallError(true)
    data class Protocol(val rpcCode: Int, val message: String) : SkillInstallError(false)
    data class VerificationFailed(val expected: String, val actual: String) : SkillInstallError(false)
    data class Unknown(val cause: String) : SkillInstallError(true)
}

File: ios/GhostCrab/Domain/Model/SkillInstallError.swift
import Foundation

/// Errors surfaced from the in-app skill install pipeline.
///
/// Each case carries an `isRetryable` flag indicating whether a retry has a
/// reasonable chance of succeeding without user intervention.
public enum SkillInstallError: Error, Sendable, Hashable {
    case unauthorized(missingScope: String)
    case notFound(slug: String)
    case dependencyConflict(conflicts: [String])
    case network(cause: String)
    case `protocol`(rpcCode: Int, message: String)
    case verificationFailed(expected: String, actual: String)
    case unknown(cause: String)

    /// Whether retrying the install is likely to help.
    public var isRetryable: Bool {
        switch self {
        case .network, .unknown:
            return true
        case .unauthorized, .notFound, .dependencyConflict, .protocol, .verificationFailed:
            return false
        }
    }
}
